File: LinearProgramming.m
% Linear Programming:
% max f(x) = 3*x1 + 5*x2
% s.t.    x1 <= 4;
%         2*2 <= 12;
%         3*x1 + 2*x2 <= 18;
%         x1 >= 0;
%         x2 >= 0;
f = [3 5];
A = [1 0; 0 2; 3 2];
b = [4; 12; 18];
Aeq = [];
beq = [];
lb = [0; 0];
ub = [];
[x val] = linprog(-f, A, b, Aeq, beq, lb, ub)
% f'x such that:
%     A * x <= b;
%     Aeq * x = beq
%     lb <= x <= ub;

t = 0:0.1:20;
% z = 0:0.1:20;
% y = [t t] .* f';
y1 = 9 - 1.5.*(t);
% for z = 0:0.1:20
%     y2 = 0.2.*z - 0.6.*(t);
%     plot(t, y2);
% end
plot(t, y1);

File: template/operationalResearch/LinearProgramming.py
from scipy.optimize import linprog
# Linear Programming solves the minimizing problem.
# https://zhuanlan.zhihu.com/p/446462562
'''
maximize: z = x + 2*y
such that:
	2*x + y <= 20
	-4*x + 5*y <= 10
	-x + 2*y >= -2
	-x + 5*y >= 15
	x >= 0
	y >= 0

'''

obj = [-1, -2] # which is minimizing -z = -x - 2*y

lhs_ineq = [[2, 1], 
	   [-4, 5], 
	   [1, -2]]

rhs_ineq = [20, 
	    10, 
	    2]

lhs_eq = [[-1, 5]]
rhs_eq = [15]
bnd = [(0, float("inf")), (0, float("inf"))] # 0 to infinitive

opt = linprog(c = obj, A_ub = lhs_ineq, b_ub = rhs_ineq,
		A_eq = lhs_eq, b_eq = rhs_eq, bounds = bnd,
		method = "revised simplex")

print(opt)

'''
参数c是指来自目标函数的系数。A_ub和b_ub分别与不等式约束左边和右边的系数有关。同样，A_eq并b_eq参考等式约束。您可以使用bounds提供决策变量的下限和上限。

您可以使用该参数method来定义要使用的线性规划方法。有以下三种选择：

    method="interior-point"选择内点法。默认情况下设置此选项。
    method="revised simplex" 选择修正的两相单纯形法。
    method="simplex" 选择传统的两相单纯形方法。

linprog() 返回具有以下属性的数据结构：

    .con 是等式约束残差。
    .fun 是最优的目标函数值（如果找到）。
    .message 是解决方案的状态。
    .nit 是完成计算所需的迭代次数。
    .slack 是松弛变量的值，或约束左右两侧的值之间的差异。
    .status是一个介于0和之间的整数4，表示解决方案的状态，例如0找到最佳解决方案的时间。
    .success是一个布尔值，显示是否已找到最佳解决方案。
    .x 是一个保存决策变量最优值的 NumPy 数组。
'''
